File: 986-interval-list-intersections/986-interval-list-intersections.java
class Solution {
    public int[][] intervalIntersection(int[][] fl, int[][] sl) {
        Arrays.sort(fl, (a, b) -> a[0] - b[0]);
        Arrays.sort(sl, (a, b) -> a[0] - b[0]);
        List<int[]> res = new ArrayList<>();
        int i = 0, j = 0;
        while(i < fl.length && j < sl.length) {
            int start = Math.max(fl[i][0], sl[j][0]);
            int end = Math.min(fl[i][1], sl[j][1]);
            if(end - start >= 0) {
                res.add(new int[]{start, end});
            }
            if(fl[i][1] < sl[j][1]) i++;
            else j++;
        }
        return res.toArray(new int[res.size()][2]);
    }
}

File: 986-interval-list-intersections/986-interval-list-intersections.py
class Solution:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:
        res = []
        firstList.sort(key=lambda x:x[0])
        secondList.sort(key=lambda x:x[0])
        m, n, p1, p2 = len(firstList), len(secondList), 0, 0
        
        while p1 < m and p2 < n:
            s1,s2 = firstList[p1], secondList[p2]
            left = max(s1[0], s2[0])
            right = min(s1[1], s2[1])
            if right - left >= 0:
                res.append([left, right])
            if s1[1] < s2[1]:
                p1 +=1
            else:
                p2 +=1
        return res
                